### Dependencies and Setup

In [1]:
import pandas as pd

import numpy as np

### File to Load

In [2]:
school_data_to_load = "Resources/schools_complete.csv"

student_data_to_load = "Resources/students_complete.csv"

### Read School and Student Data File and store into Pandas Data Frames

In [3]:
school_data = pd.read_csv(school_data_to_load, encoding='utf-8')

student_data = pd.read_csv(student_data_to_load, encoding='utf-8')

### Combine the data into a single dataset and display a preview of the dataframe

In [4]:
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])

school_data_complete.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

## District Summary Table Calculations:

#### Code will aggregate all schools in school district and return the following calculations:

* Total number of schools in district
* Total number of students in district
* Total school budget for entire district
* Average math score for district
* Average reading score for district
* Overall passing score for district
* Percent of students passing math in district
* Percent of students passing reading in district

In [5]:
total_schools = school_data_complete["school_name"].nunique()

total_students = school_data_complete["School ID"].count()

total_budget = float(school_data_complete["budget"].sum()/1000000)

avg_math = school_data_complete["math_score"].astype(float).mean()

avg_reading = school_data_complete["reading_score"].astype(float).mean()

avg_overall = float((avg_math + avg_reading)/2)

total_math_pass = (school_data_complete["math_score"].astype(float) >=70).sum()

percent_math_pass = (total_math_pass/total_students)* 100

total_read_pass = (school_data_complete["reading_score"].astype(float) >=70).sum()

percent_read_pass = (total_read_pass/total_students).astype(float) * 100



### Code adds formatting to variables derived above

In [6]:
total_students ="{:,}".format(total_students)

total_budget = "${:,.0f}".format(total_budget)

avg_math = "{:.2f}".format(avg_math)

avg_reading = "{:.2f}".format(avg_reading)

avg_overall = "{:.2f}".format(avg_overall)

percent_math_pass = "{:.2f}".format(percent_math_pass)

percent_read_pass = "{:.2f}".format(percent_read_pass)


### Code creates a data frame to hold formatted results and displays District School Summary Table

In [7]:
summary_df = pd.DataFrame({"Total Number of Schools": [total_schools],
                           
                           "Total Number of Students": [total_students],
                           
                           "Total Budget (in Millions)": [total_budget],
                           
                           "Average Math Score": [avg_math],
                           
                           "Average Reading Score": [avg_reading],
                           
                           "Overall Average Score": [avg_overall],
                           
                           "% Students Passing Math": [percent_math_pass],
                           
                           "% Students Passing Reading": [percent_read_pass]                                                    
                            })

summary_df

Total Number of Schools Total Number of Students  \
0                       15                   39,170   

  Total Budget (in Millions) Average Math Score Average Reading Score  \
0                    $82,932              78.99                 81.88   

  Overall Average Score % Students Passing Math % Students Passing Reading  
0                 80.43                   74.98                      85.81

## School Summary Table Calculations

### Code will index schools by school name and return score statistics per school. Code will return the following calculations:

* School Type 
* Total number of students per school
* Total school budget per school
* Average math score per school
* Average reading score per school
* Overall passing score per school
* Percent of students passing math per school
* Percent of students passing reading per school


In [8]:
school_name = school_data_complete["school_name"].unique()

per_school_count = school_data_complete["school_name"].value_counts()

per_school_budget = round(school_data_complete.groupby(["school_name"]).mean()["budget"],0)

per_student_budget =per_school_budget/per_school_count

per_school_math = round(school_data_complete.groupby(["school_name"]).mean()["math_score"],2)

per_school_read =  round(school_data_complete.groupby(["school_name"]).mean()["reading_score"],2)

school_pass_math =  school_data_complete[(school_data_complete["math_score"] >= 70)]

school_pass_read =  school_data_complete[(school_data_complete["reading_score"] >= 70)]

percent_school_pass_math = round(school_pass_math.groupby(["school_name"]).count()["student_name"]/per_school_count * 100,2)

percent_school_pass_read = round(school_pass_read.groupby(["school_name"]).count()["student_name"]/per_school_count * 100,2)

school_overall_pass = round((percent_school_pass_math + percent_school_pass_read)/2,2)

school_types = school_data.set_index(["school_name"])["type"]

### Code creates a data frame to store calculation results above.

In [9]:
school_summary = pd.DataFrame({"School Type": school_types,
                               
                               "Total Stuents":per_school_count,
                               
                               "Total Budget": per_school_budget,
                               
                               "Per Student Budget": per_student_budget, 
                               
                               "Average Math Score": per_school_math,
                               
                               "Average Reading Score": per_school_read,
                               
                               "% Students Passing Math": percent_school_pass_math,
                               
                               "% Students Passing Reading": percent_school_pass_read,
                               
                               "Overall Average Score": school_overall_pass                         
                                })

### Code adds formatting to variables derived above and displays the School Summary Table

In [10]:
school_summary["Total Budget"] = school_summary["Total Budget"].map("${:,.0f}".format)

school_summary["Per Student Budget"] = school_summary["Per Student Budget"].map("${:,.0f}".format)

school_summary["Total Stuents"] = school_summary["Total Stuents"].map("{:,}".format)

school_summary


School Type Total Stuents Total Budget  \
Bailey High School       District         4,976   $3,124,928   
Cabrera High School       Charter         1,858   $1,081,356   
Figueroa High School     District         2,949   $1,884,411   
Ford High School         District         2,739   $1,763,916   
Griffin High School       Charter         1,468     $917,500   
Hernandez High School    District         4,635   $3,022,020   
Holden High School        Charter           427     $248,087   
Huang High School        District         2,917   $1,910,635   
Johnson High School      District         4,761   $3,094,650   
Pena High School          Charter           962     $585,858   
Rodriguez High School    District         3,999   $2,547,363   
Shelton High School       Charter         1,761   $1,056,600   
Thomas High School        Charter         1,635   $1,043,130   
Wilson High School        Charter         2,283   $1,319,574   
Wright High School        Charter         1,800   $1,049,400   

                      Per Student Budget  Average Math Score  \
Bailey High School                  $628               77.05   
Cabrera High School                 $582               83.06   
Figueroa High School                $639               76.71   
Ford High School                    $644               77.10   
Griffin High School                 $625               83.35   
Hernandez High School               $652               77.29   
Holden High School                  $581               83.80   
Huang High School                   $655               76.63   
Johnson High School                 $650               77.07   
Pena High School                    $609               83.84   
Rodriguez High School               $637               76.84   
Shelton High School                 $600               83.36   
Thomas High School                  $638               83.42   
Wilson High School                  $578               83.27   
Wright High School                  $583               83.68   

                       Average Reading Score  % Students Passing Math  \
Bailey High School                     81.03                    66.68   
Cabrera High School                    83.98                    94.13   
Figueroa High School                   81.16                    65.99   
Ford High School                       80.75                    68.31   
Griffin High School                    83.82                    93.39   
Hernandez High School                  80.93                    66.75   
Holden High School                     83.81                    92.51   
Huang High School                      81.18                    65.68   
Johnson High School                    80.97                    66.06   
Pena High School                       84.04                    94.59   
Rodriguez High School                  80.74                    66.37   
Shelton High School                    83.73                    93.87   
Thomas High School                     83.85                    93.27   
Wilson High School                     83.99                    93.87   
Wright High School                     83.96                    93.33   

                       % Students Passing Reading  Overall Average Score  
Bailey High School                          81.93                  74.31  
Cabrera High School                         97.04                  95.58  
Figueroa High School                        80.74                  73.36  
Ford High School                            79.30                  73.81  
Griffin High School                         97.14                  95.26  
Hernandez High School                       80.86                  73.81  
Holden High School                          96.25                  94.38  
Huang High School                           81.32                  73.50  
Johnson High School                         81.22                  73.64  
Pena High School                            95.95    

## Top Performing Schools (By Passing Rate)

* Sorts and displays the top five schools in overall passing rate

In [11]:
top_school = school_summary.sort_values(["Overall Average Score"], ascending =False)

top_school.head(5)

School Type Total Stuents Total Budget Per Student Budget  \
Cabrera High School     Charter         1,858   $1,081,356               $582   
Thomas High School      Charter         1,635   $1,043,130               $638   
Pena High School        Charter           962     $585,858               $609   
Griffin High School     Charter         1,468     $917,500               $625   
Wilson High School      Charter         2,283   $1,319,574               $578   

                     Average Math Score  Average Reading Score  \
Cabrera High School               83.06                  83.98   
Thomas High School                83.42                  83.85   
Pena High School                  83.84                  84.04   
Griffin High School               83.35                  83.82   
Wilson High School                83.27                  83.99   

                     % Students Passing Math  % Students Passing Reading  \
Cabrera High School                    94.13                       97.04   
Thomas High School                     93.27                       97.31   
Pena High School                       94.59                       95.95   
Griffin High School                    93.39                       97.14   
Wilson High School                     93.87                       96.54   

                     Overall Average Score  
Cabrera High School                  95.58  
Thomas High School                   95.29  
Pena High School                     95.27  
Griffin High School                  95.26  
Wilson High School                   95.21

## Bottom Performing Schools (By Passing Rate)

* Sorts and displays the five worst-performing schools

In [12]:
bottom_school = school_summary.sort_values(["Overall Average Score"], ascending =True)

bottom_school.head(5)

School Type Total Stuents Total Budget  \
Rodriguez High School    District         3,999   $2,547,363   
Figueroa High School     District         2,949   $1,884,411   
Huang High School        District         2,917   $1,910,635   
Johnson High School      District         4,761   $3,094,650   
Ford High School         District         2,739   $1,763,916   

                      Per Student Budget  Average Math Score  \
Rodriguez High School               $637               76.84   
Figueroa High School                $639               76.71   
Huang High School                   $655               76.63   
Johnson High School                 $650               77.07   
Ford High School                    $644               77.10   

                       Average Reading Score  % Students Passing Math  \
Rodriguez High School                  80.74                    66.37   
Figueroa High School                   81.16                    65.99   
Huang High School                      81.18                    65.68   
Johnson High School                    80.97                    66.06   
Ford High School                       80.75                    68.31   

                       % Students Passing Reading  Overall Average Score  
Rodriguez High School                       80.22                  73.30  
Figueroa High School                        80.74                  73.36  
Huang High School                           81.32                  73.50  
Johnson High School                         81.22                  73.64  
Ford High School                            79.30                  73.81

## Math Scores by Grade
### Code calculates the average Math Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

In [13]:
grade_9 = round(school_data_complete[school_data_complete["grade"] == "9th"].groupby(["school_name"]).mean()["math_score"],2)

grade_10 = round(school_data_complete[school_data_complete["grade"] == "10th"].groupby(["school_name"]).mean()["math_score"],2)

grade_11 = round(school_data_complete[school_data_complete["grade"] == "11th"].groupby(["school_name"]).mean()["math_score"],2)

grade_12 = round(school_data_complete[school_data_complete["grade"] == "12th"].groupby(["school_name"]).mean()["math_score"],2)

### Code stores calculations above in dataframe and displays table

In [14]:
math_summary = pd.DataFrame({ "9th":grade_9,
                             
                              "10th": grade_10,
                             
                              "11th": grade_11,
                             
                              "12th": grade_12  
                            })

math_summary.set_index(school_name)


9th   10th   11th   12th
Huang High School      77.08  77.00  77.52  76.49
Figueroa High School   83.09  83.15  82.77  83.28
Shelton High School    76.40  76.54  76.88  77.15
Hernandez High School  77.36  77.67  76.92  76.18
Griffin High School    82.04  84.23  83.84  83.36
Wilson High School     77.44  77.34  77.14  77.19
Cabrera High School    83.79  83.43  85.00  82.86
Bailey High School     77.03  75.91  76.45  77.23
Holden High School     77.19  76.69  77.49  76.86
Pena High School       83.63  83.37  84.33  84.12
Wright High School     76.86  76.61  76.40  77.69
Rodriguez High School  83.42  82.92  83.38  83.78
Johnson High School    83.59  83.09  83.50  83.50
Ford High School       83.09  83.72  83.20  83.04
Thomas High School     83.26  84.01  83.84  83.64

## Reading Score by Grade 

* Performs the same operations as above for reading scores

In [15]:
grade_9 = round(school_data_complete[school_data_complete["grade"] == "9th"].groupby(["school_name"]).mean()["reading_score"],2)

grade_10 = round(school_data_complete[school_data_complete["grade"] == "10th"].groupby(["school_name"]).mean()["reading_score"],2)

grade_11 = round(school_data_complete[school_data_complete["grade"] == "11th"].groupby(["school_name"]).mean()["reading_score"],2)

grade_12 = round(school_data_complete[school_data_complete["grade"] == "12th"].groupby(["school_name"]).mean()["reading_score"],2)

### Code stores calculations above in dataframe and displays table

In [16]:
read_summary = pd.DataFrame({ "9th":grade_9,
                             
                             "10th": grade_10,
                             
                             "11th": grade_11,
                             
                             "12th": grade_12  
                            })

read_summary.set_index(school_name)


9th   10th   11th   12th
Huang High School      81.30  80.91  80.95  80.91
Figueroa High School   83.68  84.25  83.79  84.29
Shelton High School    81.20  81.41  80.64  81.38
Hernandez High School  80.63  81.26  80.40  80.66
Griffin High School    83.37  83.71  84.29  84.01
Wilson High School     80.87  80.66  81.40  80.86
Cabrera High School    83.68  83.32  83.82  84.70
Bailey High School     81.29  81.51  81.42  80.31
Holden High School     81.26  80.77  80.62  81.23
Pena High School       83.81  83.61  84.34  84.59
Wright High School     80.99  80.63  80.86  80.38
Rodriguez High School  84.12  83.44  84.37  82.78
Johnson High School    83.73  84.25  83.59  83.83
Ford High School       83.94  84.02  83.76  84.32
Thomas High School     83.83  83.81  84.16  84.07

## Scores by School Spending

* Creates a table that breaks down school performances based on average Spending Ranges (Per Student). 

### Code creates bins and group names to be used as lables

In [17]:
bins = [0, 585, 615, 645, 675]

group_names = ["<$585", "$585-615", "$615-645", "$645-675"]

### Code slices data by Spending Ranges

In [18]:
school_summary["Spending Ranges (Per Student)"] = pd.cut(per_student_budget, bins, labels=group_names)

spending_math_scores = round(school_summary.groupby(["Spending Ranges (Per Student)"]).mean()["Average Math Score"],2)

spending_reading_scores = round(school_summary.groupby(["Spending Ranges (Per Student)"]).mean()["Average Reading Score"],2)

spending_passing_math_scores = round(school_summary.groupby(["Spending Ranges (Per Student)"]).mean()["% Students Passing Math"],2)

spending_passing_reading_scores = round(school_summary.groupby(["Spending Ranges (Per Student)"]).mean()["% Students Passing Reading"],2)

overall_passing_rate = round((spending_passing_math_scores + spending_passing_reading_scores) / 2,2)


### Code stores calculations above in dataframe and displays table

In [19]:
spending_summary = pd.DataFrame({"Average Math Score" : spending_math_scores,
                                 
                                 "Average Reading Score": spending_reading_scores,
                                 
                                 "% Passing Math": spending_passing_math_scores,
                                 
                                 "% Passing Reading": spending_passing_reading_scores,
                                 
                                 "% Overall Passing Rate": overall_passing_rate
                                })
spending_summary


Average Math Score  Average Reading Score  \
Spending Ranges (Per Student)                                              
<$585                                       83.45                  83.94   
$585-615                                    83.60                  83.88   
$615-645                                    79.08                  81.89   
$645-675                                    77.00                  81.03   

                               % Passing Math  % Passing Reading  \
Spending Ranges (Per Student)                                      
<$585                                   93.46              96.61   
$585-615                                94.23              95.90   
$615-645                                75.67              86.11   
$645-675                                66.16              81.13   

                               % Overall Passing Rate  
Spending Ranges (Per Student)                          
<$585                                           95.04  
$585-615                                        95.06  
$615-645                                        80.89  
$645-675                                        73.64

## Scores by School Size

* Performs the same operations as above, based on school size.

### Code creates bins and group names to be used as lables

In [20]:
bins = [0, 1000, 2000, 5000]

group_names = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

### Code slices data by School Size

In [21]:
school_summary["School Size"] = pd.cut(per_school_count, bins, labels=group_names)

size_math_scores = round(school_summary.groupby(["School Size"]).mean()["Average Math Score"],2)

size_reading_scores = round(school_summary.groupby(["School Size"]).mean()["Average Reading Score"],2)

size_passing_math_scores = round(school_summary.groupby(["School Size"]).mean()["% Students Passing Math"],2)

size_passing_reading_scores = round(school_summary.groupby(["School Size"]).mean()["% Students Passing Reading"],2)

size_overall_passing_rate = round((size_passing_math_scores + size_passing_reading_scores) / 2,2)


### Code stores calculations above in dataframe and displays table

In [22]:
size_summary = pd.DataFrame({"Average Math Score" : size_math_scores, 
                             
                                 "Average Reading Score": size_reading_scores,
                             
                                 "% Passing Math": size_passing_math_scores,
                             
                                 "% Passing Reading": size_passing_reading_scores,
                             
                                 "% Overall Passing Rate": size_overall_passing_rate
                            })
size_summary

Average Math Score  Average Reading Score  % Passing Math  \
School Size                                                                     
Small (<1000)                    83.82                  83.93           93.55   
Medium (1000-2000)               83.37                  83.87           93.60   
Large (2000-5000)                77.74                  81.34           69.96   

                    % Passing Reading  % Overall Passing Rate  
School Size                                                    
Small (<1000)                   96.10                   94.82  
Medium (1000-2000)              96.79                   95.20  
Large (2000-5000)               82.77                   76.36

## Scores by School Type

* Performs the same operations as above, based on school type.

### Code slices data by School Type

In [23]:
type_math_scores = round(school_summary.groupby(["School Type"]).mean()["Average Math Score"],2)

type_reading_scores = round(school_summary.groupby(["School Type"]).mean()["Average Reading Score"],2)

type_passing_math_scores = round(school_summary.groupby(["School Type"]).mean()["% Students Passing Math"],2)

type_passing_reading_scores = round(school_summary.groupby(["School Type"]).mean()["% Students Passing Reading"],2)

type_overall_passing_rate = round((type_passing_math_scores + type_passing_reading_scores) / 2,2)

### Code stores calculations above in dataframe and displays table

In [24]:
type_summary = pd.DataFrame({    "Average Math Score" : type_math_scores,
                             
                                 "Average Reading Score": type_reading_scores,
                             
                                 "% Passing Math": type_passing_math_scores,
                             
                                 "% Passing Reading": type_passing_reading_scores,
                             
                                 "% Overall Passing Rate": type_overall_passing_rate
                            })
type_summary

Average Math Score  Average Reading Score  % Passing Math  \
School Type                                                              
Charter                   83.47                  83.90           93.62   
District                  76.96                  80.97           66.55   

             % Passing Reading  % Overall Passing Rate  
School Type                                             
Charter                  96.59                   95.10  
District                 80.80                   73.68

## Conclusion:

### According to this study, schools with higher budgets did not yield better test results. By contrast, schools with higher spending per student actually underperformed compared to schools with smaller budgets

### Additionally, smaller and medium sized schools dramatically out-performed large sized schools on passing math performances (89-91% passing vs 67%).

### Lastly, charter schools out-performed the public district schools across all metrics. However, more analysis will be required to glean if the effect is due to school practices or the fact that charter schools tend to serve smaller student populations per school. 
